In [1]:
import tensorflow as tf
from tensorflow import keras

import numpy as np
import matplotlib.pyplot as plt

In [2]:
fashion_mnist = keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

4423680/4422102 [==============================] - 1s 0us/step


In [3]:
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

In [4]:
train_images = train_images / 255.0

test_images = test_images / 255.0

Попрообуем поварьировать кол-во нейронов и слоев для улучшения точности.

In [5]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10)
])

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


In [6]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [9]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
60000/60000 [==============================] - 2s 41us/sample - loss: 0.2844 - acc: 0.8945
Epoch 2/10
60000/60000 [==============================] - 2s 41us/sample - loss: 0.2715 - acc: 0.8984
Epoch 3/10
60000/60000 [==============================] - 2s 41us/sample - loss: 0.2585 - acc: 0.9030
Epoch 4/10
60000/60000 [==============================] - 2s 39us/sample - loss: 0.2480 - acc: 0.9059
Epoch 5/10
60000/60000 [==============================] - 3s 44us/sample - loss: 0.2400 - acc: 0.9101
Epoch 6/10
60000/60000 [==============================] - 2s 41us/sample - loss: 0.2283 - acc: 0.9136
Epoch 7/10
60000/60000 [==============================] - 3s 48us/sample - loss: 0.2213 - acc: 0.9164
Epoch 8/10
60000/60000 [==============================] - 3s 47us/sample - loss: 0.2140 - acc: 0.9195
Epoch 9/10
60000/60000 [==============================] - 3s 47us/sample - loss: 0.2047 - acc: 0.9219
Epoch 10/10
60000/60000 [==============================] - 3s 43us/sample - loss: 

In [10]:
model.evaluate(test_images, test_labels)

10000/10000 [==============================] - 0s 19us/sample - loss: 0.3708 - acc: 0.8845


[0.3707964717507362, 0.8845]

In [11]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

model.evaluate(test_images, test_labels)

Epoch 1/10
60000/60000 [==============================] - 4s 65us/sample - loss: 0.4761 - acc: 0.8281
Epoch 2/10
60000/60000 [==============================] - 4s 65us/sample - loss: 0.3628 - acc: 0.8671
Epoch 3/10
60000/60000 [==============================] - 4s 63us/sample - loss: 0.3254 - acc: 0.8804
Epoch 4/10
60000/60000 [==============================] - 4s 60us/sample - loss: 0.3051 - acc: 0.8867
Epoch 5/10
60000/60000 [==============================] - 4s 61us/sample - loss: 0.2842 - acc: 0.8933
Epoch 6/10
60000/60000 [==============================] - 4s 66us/sample - loss: 0.2707 - acc: 0.8989
Epoch 7/10
60000/60000 [==============================] - 4s 65us/sample - loss: 0.2574 - acc: 0.9028
Epoch 8/10
60000/60000 [==============================] - 4s 63us/sample - loss: 0.2456 - acc: 0.9072
Epoch 9/10
60000/60000 [==============================] - 4s 62us/sample - loss: 0.2368 - acc: 0.9106
Epoch 10/10
10000/10000 [==============================] - 0s 40us/sample - loss: 

[0.3412462587714195, 0.8863]

In [12]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=8)

model.evaluate(test_images, test_labels)

Epoch 1/8
60000/60000 [==============================] - 4s 72us/sample - loss: 0.4845 - acc: 0.8219
Epoch 2/8
60000/60000 [==============================] - 5s 79us/sample - loss: 0.3618 - acc: 0.8672
Epoch 3/8
60000/60000 [==============================] - 5s 85us/sample - loss: 0.3283 - acc: 0.8781
Epoch 4/8
60000/60000 [==============================] - 5s 77us/sample - loss: 0.3069 - acc: 0.8868
Epoch 5/8
60000/60000 [==============================] - 5s 85us/sample - loss: 0.2878 - acc: 0.8929
Epoch 6/8
60000/60000 [==============================] - 5s 82us/sample - loss: 0.2736 - acc: 0.8977
Epoch 7/8
60000/60000 [==============================] - 5s 78us/sample - loss: 0.2596 - acc: 0.9014
Epoch 8/8
10000/10000 [==============================] - 0s 38us/sample - loss: 0.3467 - acc: 0.8780


[0.34666734461784365, 0.878]

In [13]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

model.evaluate(test_images, test_labels)

Epoch 1/10
60000/60000 [==============================] - 5s 79us/sample - loss: 0.4780 - acc: 0.8257
Epoch 2/10
60000/60000 [==============================] - 4s 74us/sample - loss: 0.3660 - acc: 0.8662
Epoch 3/10
60000/60000 [==============================] - 5s 78us/sample - loss: 0.3284 - acc: 0.8787
Epoch 4/10
60000/60000 [==============================] - 4s 74us/sample - loss: 0.3065 - acc: 0.8869
Epoch 5/10
60000/60000 [==============================] - 4s 73us/sample - loss: 0.2851 - acc: 0.8938
Epoch 6/10
60000/60000 [==============================] - 4s 74us/sample - loss: 0.2716 - acc: 0.8979
Epoch 7/10
60000/60000 [==============================] - 5s 80us/sample - loss: 0.2605 - acc: 0.9021
Epoch 8/10
60000/60000 [==============================] - 4s 73us/sample - loss: 0.2502 - acc: 0.9058
Epoch 9/10
60000/60000 [==============================] - 4s 74us/sample - loss: 0.2404 - acc: 0.9088
Epoch 10/10
10000/10000 [==============================] - 0s 44us/sample - loss: 

[0.35736971286535263, 0.8827]

In [14]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 6s 97us/sample - loss: 0.4918 - acc: 0.8196
Epoch 2/12
60000/60000 [==============================] - 5s 89us/sample - loss: 0.3720 - acc: 0.8634
Epoch 3/12
60000/60000 [==============================] - 4s 65us/sample - loss: 0.3371 - acc: 0.8761
Epoch 4/12
60000/60000 [==============================] - 4s 69us/sample - loss: 0.3132 - acc: 0.8830
Epoch 5/12
60000/60000 [==============================] - 4s 66us/sample - loss: 0.2947 - acc: 0.8906
Epoch 6/12
60000/60000 [==============================] - 5s 85us/sample - loss: 0.2816 - acc: 0.8942
Epoch 7/12
60000/60000 [==============================] - 5s 81us/sample - loss: 0.2662 - acc: 0.8989
Epoch 8/12
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2571 - acc: 0.9031
Epoch 9/12
60000/60000 [==============================] - 4s 65us/sample - loss: 0.2457 - acc: 0.9067
Epoch 10/12
60000/60000 [==============================] - 4s 63us/sample - loss: 

[0.33520655923485754, 0.89]

In [15]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(220, activation='relu'),
    keras.layers.Dense(220, activation='relu'),
    keras.layers.Dense(220, activation='relu'),
    keras.layers.Dense(220, activation='relu'),
    keras.layers.Dense(220, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 6s 107us/sample - loss: 0.5059 - acc: 0.8147
Epoch 2/12
60000/60000 [==============================] - 7s 110us/sample - loss: 0.3845 - acc: 0.8598
Epoch 3/12
60000/60000 [==============================] - 6s 107us/sample - loss: 0.3444 - acc: 0.8742
Epoch 4/12
60000/60000 [==============================] - 6s 106us/sample - loss: 0.3215 - acc: 0.8832
Epoch 5/12
60000/60000 [==============================] - 7s 115us/sample - loss: 0.3019 - acc: 0.8890
Epoch 6/12
60000/60000 [==============================] - 6s 104us/sample - loss: 0.2907 - acc: 0.8928
Epoch 7/12
60000/60000 [==============================] - 6s 107us/sample - loss: 0.2786 - acc: 0.8991
Epoch 8/12
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2653 - acc: 0.9019
Epoch 9/12
60000/60000 [==============================] - 5s 76us/sample - loss: 0.2561 - acc: 0.9050
Epoch 10/12
60000/60000 [==============================] - 5s 77us/sample -

[0.33898901295661926, 0.8847]

In [16]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(180, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 5s 79us/sample - loss: 0.5076 - acc: 0.8158
Epoch 2/12
60000/60000 [==============================] - 5s 80us/sample - loss: 0.3783 - acc: 0.8626
Epoch 3/12
60000/60000 [==============================] - 4s 73us/sample - loss: 0.3483 - acc: 0.8731
Epoch 4/12
60000/60000 [==============================] - 5s 78us/sample - loss: 0.3231 - acc: 0.8816
Epoch 5/12
60000/60000 [==============================] - 4s 74us/sample - loss: 0.3038 - acc: 0.8881
Epoch 6/12
60000/60000 [==============================] - 4s 75us/sample - loss: 0.2907 - acc: 0.8937
Epoch 7/12
60000/60000 [==============================] - 5s 80us/sample - loss: 0.2779 - acc: 0.8978
Epoch 8/12
60000/60000 [==============================] - 5s 76us/sample - loss: 0.2659 - acc: 0.9016
Epoch 9/12
60000/60000 [==============================] - 5s 78us/sample - loss: 0.2570 - acc: 0.9040
Epoch 10/12
60000/60000 [==============================] - 5s 87us/sample - loss: 

[0.3274161754488945, 0.8845]

In [17]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 6s 101us/sample - loss: 0.5014 - acc: 0.8179
Epoch 2/12
60000/60000 [==============================] - 6s 101us/sample - loss: 0.3805 - acc: 0.8617
Epoch 3/12
60000/60000 [==============================] - 6s 97us/sample - loss: 0.3450 - acc: 0.8741
Epoch 4/12
60000/60000 [==============================] - 6s 102us/sample - loss: 0.3225 - acc: 0.8825
Epoch 5/12
60000/60000 [==============================] - 6s 92us/sample - loss: 0.3026 - acc: 0.8887
Epoch 6/12
60000/60000 [==============================] - 6s 98us/sample - loss: 0.2907 - acc: 0.8939
Epoch 7/12
60000/60000 [==============================] - 6s 100us/sample - loss: 0.2757 - acc: 0.8975
Epoch 8/12
60000/60000 [==============================] - 6s 97us/sample - loss: 0.2639 - acc: 0.9016
Epoch 9/12
60000/60000 [==============================] - 6s 94us/sample - loss: 0.2566 - acc: 0.9041
Epoch 10/12
60000/60000 [==============================] - 6s 100us/sample - l

[0.3635029647290707, 0.8861]

Есть разница в точности между тренировочными данными и валидационной выборкой из-за переобучения модели. Попробуем добавить регуляризацию для уменьшения этой разницы.

In [18]:
from tensorflow.keras import regularizers

In [20]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=15)

model.evaluate(test_images, test_labels)

Epoch 1/15
60000/60000 [==============================] - 9s 145us/sample - loss: 0.9439 - acc: 0.8064
Epoch 2/15
60000/60000 [==============================] - 8s 137us/sample - loss: 0.5963 - acc: 0.8448
Epoch 3/15
60000/60000 [==============================] - 8s 139us/sample - loss: 0.5302 - acc: 0.8546
Epoch 4/15
60000/60000 [==============================] - 8s 126us/sample - loss: 0.5037 - acc: 0.8608
Epoch 5/15
60000/60000 [==============================] - 8s 130us/sample - loss: 0.4860 - acc: 0.8640
Epoch 6/15
60000/60000 [==============================] - 8s 126us/sample - loss: 0.4733 - acc: 0.8654
Epoch 7/15
60000/60000 [==============================] - 8s 127us/sample - loss: 0.4642 - acc: 0.8687
Epoch 8/15
60000/60000 [==============================] - 8s 126us/sample - loss: 0.4553 - acc: 0.8705
Epoch 9/15
60000/60000 [==============================] - 8s 131us/sample - loss: 0.4500 - acc: 0.8714
Epoch 10/15
60000/60000 [==============================] - 8s 128us/sampl

[0.46796150336265563, 0.8616]

In [21]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(512, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(512, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(512, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(512, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(512, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=20)

model.evaluate(test_images, test_labels)

Epoch 1/20
60000/60000 [==============================] - 14s 241us/sample - loss: 1.0462 - acc: 0.7982
Epoch 2/20
60000/60000 [==============================] - 16s 259us/sample - loss: 0.5952 - acc: 0.8420
Epoch 3/20
60000/60000 [==============================] - 16s 271us/sample - loss: 0.5367 - acc: 0.8539
Epoch 4/20
60000/60000 [==============================] - 16s 275us/sample - loss: 0.5094 - acc: 0.8600
Epoch 5/20
60000/60000 [==============================] - 17s 282us/sample - loss: 0.4939 - acc: 0.8618
Epoch 6/20
60000/60000 [==============================] - 16s 267us/sample - loss: 0.4762 - acc: 0.8662
Epoch 7/20
60000/60000 [==============================] - 16s 269us/sample - loss: 0.4646 - acc: 0.8692
Epoch 8/20
60000/60000 [==============================] - 14s 241us/sample - loss: 0.4609 - acc: 0.8690
Epoch 9/20
60000/60000 [==============================] - 14s 235us/sample - loss: 0.4562 - acc: 0.8706
Epoch 10/20
60000/60000 [==============================] - 15s 2

[0.4816193678855896, 0.8608]

In [22]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 7s 116us/sample - loss: 0.8551 - acc: 0.8124
Epoch 2/12
60000/60000 [==============================] - 7s 114us/sample - loss: 0.5786 - acc: 0.8461
Epoch 3/12
60000/60000 [==============================] - 7s 109us/sample - loss: 0.5238 - acc: 0.8536
Epoch 4/12
60000/60000 [==============================] - 7s 114us/sample - loss: 0.4925 - acc: 0.8619
Epoch 5/12
60000/60000 [==============================] - 6s 106us/sample - loss: 0.4790 - acc: 0.8642
Epoch 6/12
60000/60000 [==============================] - 7s 111us/sample - loss: 0.4663 - acc: 0.8681
Epoch 7/12
60000/60000 [==============================] - 6s 106us/sample - loss: 0.4579 - acc: 0.8683
Epoch 8/12
60000/60000 [==============================] - 7s 111us/sample - loss: 0.4490 - acc: 0.8700
Epoch 9/12
60000/60000 [==============================] - 7s 109us/sample - loss: 0.4432 - acc: 0.8725
Epoch 10/12
60000/60000 [==============================] - 6s 106us/sampl

[0.48286692638397216, 0.8551]

Регуляризация с коеффициентом 0.001 оказалась слишком сильной, попробуем ее уменьшить.

In [24]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.0001)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.0001)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.0001)),
    keras.layers.Dense(200, activation='relu', kernel_regularizer=regularizers.l2(0.0001)),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 8s 128us/sample - loss: 0.5677 - acc: 0.8189
Epoch 2/12
60000/60000 [==============================] - 7s 121us/sample - loss: 0.4441 - acc: 0.8620
Epoch 3/12
60000/60000 [==============================] - 7s 117us/sample - loss: 0.4133 - acc: 0.8720
Epoch 4/12
60000/60000 [==============================] - 7s 115us/sample - loss: 0.3923 - acc: 0.8786
Epoch 5/12
60000/60000 [==============================] - 7s 119us/sample - loss: 0.3755 - acc: 0.8840
Epoch 6/12
60000/60000 [==============================] - 7s 114us/sample - loss: 0.3639 - acc: 0.8884
Epoch 7/12
60000/60000 [==============================] - 7s 117us/sample - loss: 0.3541 - acc: 0.8900
Epoch 8/12
60000/60000 [==============================] - 7s 115us/sample - loss: 0.3485 - acc: 0.8923
Epoch 9/12
60000/60000 [==============================] - 7s 122us/sample - loss: 0.3388 - acc: 0.8956
Epoch 10/12
60000/60000 [==============================] - 7s 123us/sampl

[0.39509349257946014, 0.8818]

In [25]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0002)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0002)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0002)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0002)),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 9s 142us/sample - loss: 0.6377 - acc: 0.8201
Epoch 2/12
60000/60000 [==============================] - 8s 130us/sample - loss: 0.4929 - acc: 0.8585
Epoch 3/12
60000/60000 [==============================] - 8s 135us/sample - loss: 0.4454 - acc: 0.8696
Epoch 4/12
60000/60000 [==============================] - 8s 130us/sample - loss: 0.4196 - acc: 0.8748
Epoch 5/12
60000/60000 [==============================] - 8s 137us/sample - loss: 0.4008 - acc: 0.8787
Epoch 6/12
60000/60000 [==============================] - 8s 130us/sample - loss: 0.3917 - acc: 0.8816
Epoch 7/12
60000/60000 [==============================] - 9s 142us/sample - loss: 0.3806 - acc: 0.8852s - loss: 0.3799 - acc:
Epoch 8/12
60000/60000 [==============================] - 8s 133us/sample - loss: 0.3709 - acc: 0.8882
Epoch 9/12
60000/60000 [==============================] - 9s 147us/sample - loss: 0.3662 - acc: 0.8904
Epoch 10/12
60000/60000 [=========================

[0.44644105544090273, 0.8632]

In [26]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0003)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0003)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0003)),
    keras.layers.Dense(256, activation='relu', kernel_regularizer=regularizers.l2(0.0003)),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 9s 151us/sample - loss: 0.6910 - acc: 0.8191
Epoch 2/12
60000/60000 [==============================] - 8s 137us/sample - loss: 0.5214 - acc: 0.8553
Epoch 3/12
60000/60000 [==============================] - 9s 143us/sample - loss: 0.4645 - acc: 0.8664
Epoch 4/12
60000/60000 [==============================] - 8s 139us/sample - loss: 0.4383 - acc: 0.8727
Epoch 5/12
60000/60000 [==============================] - 9s 143us/sample - loss: 0.4176 - acc: 0.8760
Epoch 6/12
60000/60000 [==============================] - 8s 138us/sample - loss: 0.4071 - acc: 0.8798
Epoch 7/12
60000/60000 [==============================] - 9s 142us/sample - loss: 0.3941 - acc: 0.8836
Epoch 8/12
60000/60000 [==============================] - 9s 143us/sample - loss: 0.3898 - acc: 0.8842
Epoch 9/12
60000/60000 [==============================] - 9s 143us/sample - loss: 0.3803 - acc: 0.8866
Epoch 10/12
60000/60000 [==============================] - 9s 149us/sampl

[0.4354605749845505, 0.868]

Теперь попробуем другой способ регуляризации. Добавим Dropout слой.

In [29]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 8s 133us/sample - loss: 0.5219 - acc: 0.8119
Epoch 2/12
60000/60000 [==============================] - 8s 128us/sample - loss: 0.3870 - acc: 0.8607
Epoch 3/12
60000/60000 [==============================] - 7s 124us/sample - loss: 0.3513 - acc: 0.8726
Epoch 4/12
60000/60000 [==============================] - 8s 128us/sample - loss: 0.3289 - acc: 0.8807
Epoch 5/12
60000/60000 [==============================] - 7s 121us/sample - loss: 0.3098 - acc: 0.8879
Epoch 6/12
60000/60000 [==============================] - 8s 126us/sample - loss: 0.3002 - acc: 0.8903
Epoch 7/12
60000/60000 [==============================] - 7s 124us/sample - loss: 0.2881 - acc: 0.8939
Epoch 8/12
60000/60000 [==============================] - 8s 127us/sample - loss: 0.2777 - acc: 0.8972
Epoch 9/12
60000/60000 [==============================] - 8s 125us/sample - loss: 0.2712 - acc: 0.9010
Epoch 10/12
60000/60000 [==============================] - 8s 128us/sampl

[0.3370477528572082, 0.8804]

In [30]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(256, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=12)

model.evaluate(test_images, test_labels)

Epoch 1/12
60000/60000 [==============================] - 12s 194us/sample - loss: 0.5817 - acc: 0.7912
Epoch 2/12
60000/60000 [==============================] - 11s 181us/sample - loss: 0.4410 - acc: 0.8429
Epoch 3/12
60000/60000 [==============================] - 11s 177us/sample - loss: 0.4088 - acc: 0.8545
Epoch 4/12
60000/60000 [==============================] - 11s 182us/sample - loss: 0.3897 - acc: 0.8614
Epoch 5/12
60000/60000 [==============================] - 11s 183us/sample - loss: 0.3758 - acc: 0.8659
Epoch 6/12
60000/60000 [==============================] - 11s 180us/sample - loss: 0.3654 - acc: 0.8701
Epoch 7/12
60000/60000 [==============================] - 11s 182us/sample - loss: 0.3546 - acc: 0.8730
Epoch 8/12
60000/60000 [==============================] - 11s 186us/sample - loss: 0.3436 - acc: 0.8770
Epoch 9/12
60000/60000 [==============================] - 11s 186us/sample - loss: 0.3353 - acc: 0.8807
Epoch 10/12
60000/60000 [==============================] - 11s 1

[0.3502184754014015, 0.881]

In [31]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(512, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(512, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(512, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(512, activation='relu'),
    keras.layers.Dropout(0.2),
    keras.layers.Dense(512, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=14)

model.evaluate(test_images, test_labels)

Epoch 1/14
60000/60000 [==============================] - 14s 240us/sample - loss: 0.5682 - acc: 0.7948
Epoch 2/14
60000/60000 [==============================] - 14s 232us/sample - loss: 0.4473 - acc: 0.8417
Epoch 3/14
60000/60000 [==============================] - 17s 277us/sample - loss: 0.4121 - acc: 0.8546
Epoch 4/14
60000/60000 [==============================] - 16s 272us/sample - loss: 0.3927 - acc: 0.8626
Epoch 5/14
60000/60000 [==============================] - 16s 269us/sample - loss: 0.3768 - acc: 0.8658
Epoch 6/14
60000/60000 [==============================] - 15s 253us/sample - loss: 0.3659 - acc: 0.8699
Epoch 7/14
60000/60000 [==============================] - 15s 243us/sample - loss: 0.3529 - acc: 0.8747
Epoch 8/14
60000/60000 [==============================] - 16s 259us/sample - loss: 0.3485 - acc: 0.8755
Epoch 9/14
60000/60000 [==============================] - 15s 250us/sample - loss: 0.3376 - acc: 0.8816
Epoch 10/14
60000/60000 [==============================] - 15s 2

[0.3497237119436264, 0.8785]

В итоге самая лучшаая модель получилась с 10 слоями 4 из которых это Dropout, по 256 нейрона в слое. Ее точность на валидационной выборке максимально приближена к точности на тренировочной выборке и составила 0.8810.